src/components/ripple/ripple.ts

```diff
@@ -0,0 +1,288 @@
+import {
+  ChangeDetectionStrategy,
+  Component,
+  ElementRef,
+  HostBinding,
+  Input,
+  OnChanges,
+  OnDestroy,
+  OnInit,
+  Renderer,
+  SimpleChange,
+  ViewEncapsulation,
+} from '@angular/core';
+
+enum RippleState {
+  NEW,
+  EXPANDING,
+  FADING_OUT,
+}
+
+class Ripple {
+  state = RippleState.NEW;
+  constructor(public element: Element) {}
+}
+
+const RIPPLE_SPEED_PX_PER_SECOND = 500;
+const MIN_RIPPLE_FILL_TIME_SECONDS = 0.2;
+const MAX_RIPPLE_FILL_TIME_SECONDS = 0.6;
+
+const sqr = (x: number) => x * x;
+
+const distanceToFurthestCorner = (x: number, y: number, parentRect: any) => {
+  const x1 = parentRect.left;
+  const x2 = x1 + parentRect.width;
```

Comment:
Instead of adding `width` and `height`, can't you just use `bottom` and `right` from the `ClientRect`?